llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -592,8 +662,223 @@ static Error applyAddends(MCCASReader &Reader,
+    case dwarf::DW_FORM_strx: {
+      DWARFDataExtractor DWARFExtractor(toStringRef(CUData), IsLittleEndian,
+                                        AddressSize);
+      uint64_t PrevOffset = CUOffset;
+      DWARFExtractor.getULEB128(&CUOffset);
+      FormSize += CUOffset - PrevOffset;
+      break;
+    }
+    case dwarf::DW_FORM_LLVM_addrx_offset: {
+      DWARFDataExtractor DWARFExtractor(toStringRef(CUData), IsLittleEndian,
+                                        AddressSize);
+      uint64_t PrevOffset = CUOffset;
+      DWARFExtractor.getULEB128(&CUOffset);
+      FormSize += CUOffset - PrevOffset + 4;
+      break;
+    }
+    case dwarf::DW_FORM_string: {
+      DWARFDataExtractor DWARFExtractor(toStringRef(CUData), IsLittleEndian,
+                                        AddressSize);
+      auto CurrOffset = CUOffset;
+      DWARFExtractor.getCStr(&CUOffset);
+      FormSize += CUOffset - CurrOffset;
+      break;
+    }
+    case dwarf::DW_FORM_indirect: {
+      DWARFDataExtractor DWARFExtractor(toStringRef(CUData), IsLittleEndian,
+                                        AddressSize);
+      uint64_t PrevOffset = CUOffset;
+      FormVal =
+          static_cast<dwarf::Form>(DWARFExtractor.getULEB128(&CUOffset, &Err));
+      Indirect = true;
+      FormSize += CUOffset - PrevOffset;
+      break;
+    }
+    case dwarf::DW_FORM_strp:
+    case dwarf::DW_FORM_sec_offset:
+    case dwarf::DW_FORM_GNU_ref_alt:
+    case dwarf::DW_FORM_GNU_strp_alt:
+    case dwarf::DW_FORM_line_strp:
+    case dwarf::DW_FORM_strp_sup: {
+      FormSize += FP.getDwarfOffsetByteSize();
+      break;
+    }
+    case dwarf::DW_FORM_ref_sig8: {
+      FormSize += 8;
+      break;
+    }
+    case dwarf::DW_FORM_implicit_const:
+    case dwarf::DW_FORM_flag_present: {
+      FormSize += 0;
+      break;
+    }
+    default:
+      llvm_unreachable("usupported form");
+    }
+  } while (Indirect && !Err);
+
+  if (Err)
+    return Err;
+  return FormSize;
+}
+
+static void materializeCUDie(DWARFCompileUnit &DCU,
+                             SmallVector<char> &SectionContents,
+                             SmallVector<char, 40> &CUData,
+                             ArrayRef<unsigned char> FixupsDataArrayRef,
+                             uint64_t &CUOffset, uint64_t &FixupOffset,
+                             uint64_t &SectionOffset) {
+  // Copy Abbrev Tag
+  DWARFDataExtractor DWARFExtractor(toStringRef(CUData), DCU.isLittleEndian(),
+                                    DCU.getAddressByteSize());
+  uint64_t PrevOffset = CUOffset;
+  uint64_t AbbrevTag = DWARFExtractor.getULEB128(&CUOffset);
+  std::memcpy(&SectionContents[SectionOffset], &AbbrevTag,
+              CUOffset - PrevOffset);
+  SectionOffset += CUOffset - PrevOffset;
+
+  // Return if the abbrev tag is 0, this indicates the end of a sequence of
+  // children DWARFDies
+  if (AbbrevTag == 0)
+    return;
+
+  // Create a dummy DWARFDie to extract the attributes.
```

Comment:
why `dummy`?

Reply:
It is a dummy die because it doesn't belong to the `DWARFCompileUnit`, we create our own and only fill the `DWARFDebugInfoEntry` with enough metadata to get the abbreviations. Is there a better word than dummy? Incomplete maybe? The point is, the `DWARFCompileUnit` doesn't have enough metadata to create a legal Die itself and we have to create an incomplete one with enough metadata to have iterable forms

Reply:
I am fine with that